coalib/misc/Caching.py

```diff
@@ -0,0 +1,81 @@
+import calendar
+import time
+import hashlib
+
+from coalib.misc.CachingUtilities import (
+    pickle_load, pickle_dump, time_consistent, update_time_db)
+
+
+class FileCache:
+
+    def __init__(self, log_printer, project_dir, flush_cache=False):
+        """
+        Initialize cache instance for project.
+
+        :param log_printer: A LogPrinter object to use for logging.
+        :param project_dir: The root directory of the project to be used
+                            as a key identifier.
+        :param flush_cache: Flush the cache and rebuild it.
+        """
+        self.log_printer = log_printer
+        self.project_dir = project_dir.encode("utf-8")
+        self.md5sum = hashlib.md5(self.project_dir).hexdigest()
+        self.data = {}
+        if not time_consistent(log_printer, self.md5sum):
+            log_printer.warn("It seems like you went back in time - your "
+                             "system time is behind the last recorded run "
+                             "time on this project. The cache will "
+                             "be flushed and rebuilt.")
+            flush_cache = True
+        if not flush_cache:
+            self.data = pickle_load(log_printer, self.md5sum, {})
+        else:
+            log_printer.info("The file cache was successfully flushed.")
+        self.changed_files = set()
+
+    def write(self):
+        """
+        Update the last run time on the project for each file
+        to the current time.
```

Comment:
the comment isn't precise, this doesn't really update it (it updates the dict but that's only a lazy thing, right? That could've happened before in add_to_changed_files - and maybe it should have?)